pkg/registry/core/service/ipallocator/allocator.go

```diff
@@ -127,7 +127,7 @@ func New(cidr *net.IPNet, allocatorFactory allocator.AllocatorFactory) (*Range,
 // NewInMemory creates an in-memory allocator.
 func NewInMemory(cidr *net.IPNet) (*Range, error) {
 	return New(cidr, func(max int, rangeSpec string) (allocator.Interface, error) {
-		return allocator.NewAllocationMap(max, rangeSpec), nil
+		return allocator.NewAllocationReservedMap(max, rangeSpec), nil
```

Comment:
we have to do it in the bitmap because the randomly chosen ip happens in the bitmap, see L238 of this file
```
func (r *Range) allocateNext(dryRun bool) (net.IP, error) {
	label := r.CIDR()
	if dryRun {
		// Don't bother finding a free value. It's racy and not worth the
		// effort to plumb any further.
		return r.CIDR().IP, nil
	}
	offset, ok, err := r.alloc.AllocateNext()
	if err != nil {
```

the bitmap marks the ip as allocated, if we decide that we don't like it, we have to release it and call allocatenext() again until we find one in the desired range.
It is not that is just less performant, it also can race with the allocation logic, since the bitmap is snapshoted and restored from etcd, it can happen that in one of those roundtrips it has changed by other apiserver